src/kernel/emulation/linux/misc/sysctl_machdep.c

```diff
@@ -51,3 +91,164 @@ sysctl_handler(handle_brand_string)
 	return 0;
 }
 
+sysctl_handler(handle_max_basic)
+{
+	setup(0);
+
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+
+
+
+	copyout_int(eax, (char*)old,oldlen);
+
+	return 0;
+}
+
+sysctl_handler(handle_family)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 7;
+    eax &= 15;
+
+    copyout_int(eax, old, oldlen);
+
+    return 0;
+}
+
+sysctl_handler(handle_model)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 3;
+    eax &= 15;
+
+    copyout_int(eax,(char*)old, oldlen);
+
+    return 0;
+}
+
+sysctl_handler(handle_stepping)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax &= 15;
+
+    copyout_int(eax,(char*)old, oldlen);
+
+    return 0;
+}
+
+sysctl_handler(handle_brand_string)
+{
+    setup(0x80000000);
+
+    __cpuid(level,eax,ebx, ecx, edx);
+
+    if(eax < 0x80000004) // the information is not implemented
+        return 2;
+
+
+
+    union
+    {
+        unsigned int brand[12];
+        char name[49];
+    } v;
+
+		for (int i = 1; i < 4; i++)
+		{
+				eax = 0x80000000+i;
+				__cpuid(level, eax, ebx, ecx, edx);
+				v.brand[0x0+(i-1)*4] = eax;
+				v.brand[0x1+(i-1)*4] = ebx;
+				v.brand[0x2+(i-1)*4] = ecx;
+				v.brand[0x3+(i-1)*4] = edx;
+		}
+
+    v.name[48] = 0;
+    copyout_string(v.name, (char*) old, oldlen);
+
+    return 0;
+}
+sysctl_handler(handle_features)
+{
+
+    setup(1);
+
+        static const char features[][7] = {"FPU","VME", "DE", "PSE", "TSC", "MSR", "PAE", "MCE", "CX8", "APIC", "","SEP","MTRR","PGE",
+                            "MCA", "CMOV", "PAT", "PSE-36", "PSN", "CLFSH", "", "DS", "ACPI", "MMX", "FXSR", "SSE", "SSE2", "SS",
+                            "HTT", "TM","IA64","PBE"};
+
+        __cpuid(0,eax,ebx,ecx,edx);
+
+    if(old != NULL)
+    {
+
+        int counter = 0;
+        int j = 0;
+
+        for (int i = 0; i < 32; i++)
+        {
+
+            if(i == 10 || i == 20)
+                continue;
+
+            if(edx>>i&1 && counter < *oldlen)
+            {
+                int len = __simple_strlen(features[i]);
+                
+                strncpy(old,features[i]);
```

Comment:
This is `strncpy`, not `strcpy` (`strncpy` is safer, though; good choice 👍), you need a maximum length argument.

Also, you need to offset the destination pointer by the correct amount. Note that this suggestion is assuming that you move the conditional below above this `strncpy`, like I suggested. Otherwise, you would need to add 1 to your offset if a space is going to be added.
```suggestion
                // the length argument here is essentially `min(len, *oldlen - counter)`
                strncpy(outstr + counter, features[i], len < (*oldlen - counter) ? len : (*oldlen - counter));
```
Don't forget to replace the declaration of `strcpy` with:
```c
extern char *strncpy(char *dest, const char *src, __SIZE_TYPE__ n);
```